libcalico-go/lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -246,15 +329,55 @@ func (c *blockAffinityClient) Get(ctx context.Context, key model.Key, revision s
 	return v1kvp, nil
 }
 
-func (c *blockAffinityClient) List(ctx context.Context, list model.ListInterface, revision string) (*model.KVPairList, error) {
+func (c *blockAffinityClient) getV3(ctx context.Context, key model.ResourceKey, revision string) (*model.KVPair, error) {
+	kvp, err := c.rc.Get(ctx, key, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	// If this object has been marked as deleted, then we need to clean it up and
+	// return not found.
+	if kvp.Value.(*libapiv3.BlockAffinity).Spec.Deleted == "true" {
+		if _, err := c.DeleteKVP(ctx, kvp); err != nil {
+			return nil, err
+		}
+		return nil, cerrors.ErrorResourceDoesNotExist{Err: fmt.Errorf("Resource was deleted"), Identifier: key}
+	}
+
+	return kvp, nil
+}
+
+func (c *blockAffinityClient) Get(ctx context.Context, key model.Key, revision string) (*model.KVPair, error) {
+	if isV1Key(key) {
+		// If this is a V1 resource, then it is from the IPAM code.
+		// Convert it, but treat it as a V1 resource.
+		return c.getV1(ctx, key.(model.BlockAffinityKey), revision)
+	}
+	// If this is a V3 resource, then it is already in CRD format.
+	return c.getV3(ctx, key.(model.ResourceKey), revision)
+}
+
+func isV1List(list model.ListInterface) bool {
+	switch list.(type) {
+	case model.BlockAffinityListOptions:
+		return true
+	case model.ResourceListOptions:
+		return false
+	default:
+		log.Panic("blockAffinityClient : wrong key interface type")
+	}
+	return false
+}
+
+func (c *blockAffinityClient) listV1(ctx context.Context, list model.BlockAffinityListOptions, revision string) (*model.KVPairList, error) {
 	l := model.ResourceListOptions{Kind: libapiv3.KindBlockAffinity}
 	v3list, err := c.rc.List(ctx, l, revision)
 	if err != nil {
 		return nil, err
 	}
 
-	host := list.(model.BlockAffinityListOptions).Host
-	requestedIPVersion := list.(model.BlockAffinityListOptions).IPVersion
```

Comment:
Huh, I think this is a pre-existing bug... bug we include affinities with Deleted=true when listing them, but not when performing a Get(). Oh well.

Reply:
Might actually need to fix this one.